index.coffee

```diff
@@ -1,60 +1,172 @@
+		@bytesRead += data.length if data?
+		@_reportProgress()
+		return super(data, encoding)
+
+	_end: ->
+		return if @ended
+		@ended = true
+		doEnd = =>
+			@request.abort()
+			@_retry.cancel()
+			@_reportProgress.cancel()
+			@_reportProgress.apply()
+			@emit('end')
+			if not @error?
+				@emit('complete')
+		if @response?
+			@response.end(doEnd)
+		else
+			doEnd()
+
+	_read: -> # noop -- we're manually pushing buffers into self
+
+	progress: ->
+		retries: @retries
+		maxRetries: @maxRetries
+		percentage: Math.round(100 * @bytesRead / @bytesTotal) or 0
+		size: { transferred: @bytesRead, total: @bytesTotal }
+
+	_reportProgress: (progress = {}) =>
+		@emit('progress', Object.assign @progress(), progress)
 
 	# Sends the requests and sets up listeners in order
 	# to be able to resume.
 	# Returns a stream object similar to the result of `request`.
 	# 'response' and 'request' events on the stream are generated
 	# based on the first request, in order to emit them only once.
-	resume: ->
-		@requestOpts.headers ?= {}
-		@requestOpts.headers.range = "bytes=#{@bytesRead}-"
-		@requestModule(@requestOpts)
-		.on 'request', (reqObj) =>
-			if @retries is 0
-				@stream.emit('request', reqObj)
-		.on 'response', (response) =>
+	_request: ->
+		if @response?
+			# the first response is emitted to listeners, so only specify the
+			# Range header on subsequent requests, to avoid causing a 206 response
+			# status code, which may be unexpected.
+			@requestOpts.headers ?= {}
+			@requestOpts.headers.range = "bytes=#{@bytesRead}-"
+
+		initial = @request is null
+		aborted = false
+		errored = false
+
+		@request = @requestModule(@requestOpts)
+		.once 'request', (reqObj) =>
+			if initial
+				@emit('request', reqObj)
+		.once 'response', (response) =>
+			@_follow(response)
 			if response.statusCode >= 400
 				@error = new Error("Request failed with status code #{response.statusCode}")
-				@stream.emit('error', @error)
-			else if @retries is 0
-				@stream.emit('response', response)
-				@contentLength = parseInt(response.headers['content-length'])
+				@_retry() # will abort the request
 		.on 'error', (err) =>
-			@stream.emit('socketError', err)
-			@retryOrEnd()
-		.on('end', => @retryOrEnd())
-		.pipe(@stream, end: false)
+			errored = true
+			@emit('socketError', err)
+			if aborted
+				@_retry()
+		.once 'abort', ->
+			aborted = true
+		.once 'complete', =>
+			if errored or (@bytesTotal? and @bytesRead < @bytesTotal)
+				@_retry()
+			else
+				@_end() # received complete response
+
+	_follow: (response, requestId) ->
+		if not @response?
+			@bytesTotal ?= parseInt(response.headers['content-length'])
+			if response.headers['accept-ranges']?.toLowerCase() != 'bytes'
+				# set the error but don't interrupt the request in case it succeeds
+				# (the common case). on first retry, we'll pick up the error and abort
+				@error = new Error('Server does not support Byte Serving')
```

Comment:
Ok, can you hold off on adding reviewers if the PR is still in heavy flux - it'd help avoid situations like this